crmd/notify.c

```diff
@@ -20,120 +20,501 @@
+        .years = hr_dt->years,
+        .months = hr_dt->months,
+        .days = hr_dt->days,
+        .seconds = hr_dt->seconds,
+        .offset = hr_dt->offset,
+        .duration = hr_dt->duration
+    };
+}
 
-    free(notify_target);
-    notify_target = NULL;
+crm_time_hr_t *
+crm_time_timeval_hr_convert(crm_time_hr_t *target, struct timeval *tv)
+{
+    crm_time_t dt;
+    crm_time_hr_t *ret;
 
-    if(script == NULL || safe_str_eq(script, "/dev/null")) {
-        crm_notice("Notifications disabled");
+    crm_time_set_timet(&dt, &tv->tv_sec);
+    ret = crm_time_hr_convert(target, &dt);
+    if (ret) {
+        ret->useconds = tv->tv_usec;
+    }
+    return ret;
+}
+
+crm_time_hr_t *
+crm_time_hr_new(const char *date_time)
+{
+    crm_time_hr_t *hr_dt = NULL;
+    struct timeval tv_now;
+
+    if (date_time == NULL) {
+        if (gettimeofday(&tv_now, NULL) == 0) {
+            hr_dt = crm_time_timeval_hr_convert(NULL, &tv_now);
+        }
+    } else {
+        crm_time_t *dt;
+
+        dt = parse_date(date_time);
+        hr_dt = crm_time_hr_convert(NULL, dt);
+        crm_time_free(dt);
+    }
+    return hr_dt;
+}
+
+void
+crm_time_hr_free(crm_time_hr_t * hr_dt)
+{
+    if (hr_dt == NULL) {
         return;
     }
+    free(hr_dt);
+}
+
+char *
+crm_time_format_hr(const char *format, crm_time_hr_t * hr_dt)
+{
+    const char *mark_s;
+    int max = 128, scanned_pos = 0, printed_pos = 0, fmt_pos = 0,
+        date_len = 0, nano_digits, fmt_len;
+    char nano_s[10], date_s[max+1], nanofmt_s[5] = "%", *tmp_fmt_s;
+    struct tm tm;
+    crm_time_t dt;     
+
+    if (!format) {
+        return NULL;
+    }
+    crm_time_set_hr_dt(&dt, hr_dt);
+    ha_get_tm_time(&tm, &dt);
+    sprintf(nano_s, "%06d000", hr_dt->useconds);
+
+    while ((format[scanned_pos]) != '\0') {
+        fmt_len = 0; 
+        mark_s = strchr(&format[scanned_pos], '%');
+        if (mark_s) {
+            fmt_pos = mark_s - format;
+            fmt_len = 1; 
+            while ((format[fmt_pos+fmt_len] != '\0') && 
+                (format[fmt_pos+fmt_len] >= '0') && 
+                (format[fmt_pos+fmt_len] <= '9')) {
+                fmt_len++;
+            }
+            scanned_pos = fmt_pos + fmt_len + 1;
+            if (format[fmt_pos+fmt_len] == 'N') {
+                nano_digits = atoi(&format[fmt_pos+1]);
+                nano_digits = (nano_digits > 6)?6:nano_digits;
+                nano_digits = (nano_digits < 0)?0:nano_digits;
+                sprintf(&nanofmt_s[1], ".%ds", nano_digits);               
+            } else {
+                if (format[fmt_pos+fmt_len] != 0) {
+                    continue;
+                }
+                fmt_pos = scanned_pos; /* print till end */
+            }
+        } else {
+            scanned_pos = strlen(format);
+            fmt_pos = scanned_pos; /* print till end */
+        }
+        tmp_fmt_s = strndup(&format[printed_pos], fmt_pos - printed_pos);
+#pragma GCC diagnostic push
```

Comment:
gcc supports push/pop since 4.6, best to conditionalize, e.g., akin to a recent warning suppression in libqb:

- https://github.com/ClusterLabs/libqb/commit/ea35cfaa616096dd1d770fbda5a735108e1725f8
- https://github.com/ClusterLabs/libqb/commit/e4909329071e3b3805609e38c535e75f69cf9c2f

Can certainly be done in a smoother fashion (also statically per defines injected by the very compiler).